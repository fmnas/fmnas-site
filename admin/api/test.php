<?php
var_dump($_SERVER['REQUEST_METHOD'], $_GET['v'], $_POST);