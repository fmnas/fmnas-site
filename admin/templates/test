I am a template!

The pet's name is {{name}}

<script>console.log("[Did I make it through?](https://forgetmenotshelter.org)");</script>