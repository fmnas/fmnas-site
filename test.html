<!DOCTYPE html>
<title>Test file</title>
<p>Hopefully copyright-check won't find me!</p>
