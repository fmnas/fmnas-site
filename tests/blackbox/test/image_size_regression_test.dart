/*
 * Copyright 2022 Google LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import 'dart:convert';
import 'dart:io';

import 'package:json_annotation/json_annotation.dart';
import 'package:test/test.dart';

import '../bin/image_size.dart';
import '../bin/results.dart';
import '../bin/service.dart';

part 'image_size_regression_test.g.dart';

const imageDir = '../data/images';
const goldenFile = 'goldens/image_size_regression_test.json';
const parallelColumns = [1, 2, 3];

// If (new value)/(old value) goes above this threshold, fail the test.
const maxFailThreshold = 2.0;
const avgFailThreshold = 1.1;
const ramFailThreshold = 1.1;

// If (new value)/(old value) goes below this threshold, update the goldens.
const maxUpdateThreshold = 0.7;
const avgUpdateThreshold = 0.9;
const ramUpdateThreshold = 0.9;

// Averages of fewer than this many data points won't be compared.
const avgCompareThreshold = 3;
// Durations under this many milliseconds won't be compared.
const msCompareThreshold = 100;

@JsonSerializable()
class ImageSizeRegressionTestEnvResults {
  ImageSizeRegressionTestEnvResults(this.images);
  Map<String, ImageResult> images;
  factory ImageSizeRegressionTestEnvResults.fromJson(Map<String, dynamic> json) => _$ImageSizeRegressionTestEnvResultsFromJson(json);
  Map<String, dynamic> toJson() => _$ImageSizeRegressionTestEnvResultsToJson(this);
}

@JsonSerializable()
class ImageSizeRegressionTestResults {
  ImageSizeRegressionTestResults(this.environments);
  Map<String, ImageSizeRegressionTestEnvResults> environments;
  factory ImageSizeRegressionTestResults.fromJson(Map<String, dynamic> json) => _$ImageSizeRegressionTestResultsFromJson(json);
  Map<String, dynamic> toJson() => _$ImageSizeRegressionTestResultsToJson(this);
}

void main() async {
  final env = Service.environmentKey('IMAGE_SIZE_ENDPOINT');

  final golden = File('goldens/image_size_regression_test.json');
  final goldenValues = golden.existsSync() ? ImageSizeRegressionTestResults.fromJson(jsonDecode(golden.readAsStringSync())) : ImageSizeRegressionTestResults({});

  // If ok is true at the end of all tests, a the golden file will be written.
  // This will have no effect if the golden values don't change past the
  // thresholds configured above.
  var ok = true;
  goldenValues.environments[env] ??= ImageSizeRegressionTestEnvResults({});
  final Map<String, ImageResult> envGoldens = goldenValues.environments[env]!.images;

  final Map<String, ImageResult> results = {};
  await ImageSize.runBenchmark(
    endpoint: Platform.environment['IMAGE_SIZE_ENDPOINT'] ??
        ImageSize.defaultEndpoint,
    parallelColumns: parallelColumns,
    binarySearchLimit: 0,
  ).listen((result) {
    print(result);
    results[result.name] = result;
  }).asFuture();
  ImageResult.printAll(results.values);

  for (final ImageResult imageResult in results.values) {
    final goldenImageResult = envGoldens[imageResult.name] ?? ImageResult();
    goldenImageResult.name = imageResult.name;
    goldenImageResult.size = imageResult.size;
    goldenImageResult.dimensions = imageResult.dimensions;
    goldenImageResult.group = imageResult.group;

    imageResult.parallel.columns.forEach((parallelism, result) {
      // Test request success.
      if (result.failed > 0) {
        ok = false;
      }
      test('$parallelism requests for ${imageResult.name} succeeded', () =>
          expect(result.succeeded, equals(parallelism)));

      ParallelResult? goldenParallelResult = goldenImageResult.parallel
          .columns[parallelism];
      if (goldenParallelResult != null) {
        if (result.max >= msCompareThreshold) {
          int maxThreshold = (goldenParallelResult.max * maxFailThreshold)
              .round();
          test('$parallelism requests for ${imageResult
              .name} took under $maxThreshold ms', () =>
              expect(result.max, lessThanOrEqualTo(maxThreshold)));
          if (result.max > maxThreshold) {
            ok = false;
          }
          if (result.max <
              (goldenParallelResult.max * maxUpdateThreshold).round()) {
            goldenParallelResult.max = result.max;
          }
        }

        if (parallelism >= avgCompareThreshold && result.avg >= msCompareThreshold) {
          int avgThreshold = (goldenParallelResult.avg * avgFailThreshold)
              .round();
          test('$parallelism requests for ${imageResult
              .name} averaged under $avgThreshold ms', () =>
              expect(result.avg, lessThanOrEqualTo(avgThreshold)));
          if (result.avg > avgThreshold) {
            ok = false;
          }
          if (result.avg <
              (goldenParallelResult.avg * avgUpdateThreshold).round()) {
            goldenParallelResult.total = result.total;
          }
        }

        if (goldenParallelResult.ram != null && result.ram != null) {
          int ramThreshold = (goldenParallelResult.ram! * ramFailThreshold)
              .round();
          test('$parallelism requests for ${imageResult
              .name} used under ${ramThreshold ~/ 1024} MiB', () =>
              expect(result.ram, lessThanOrEqualTo(ramThreshold)));
          if (result.ram! > ramThreshold) {
            ok = false;
          }
          if (result.ram! <
              (goldenParallelResult.ram! * ramUpdateThreshold).round()) {
            goldenParallelResult.ram = result.ram;
          }
        } else {
          goldenParallelResult.ram = result.ram;
        }
      } else {
        goldenParallelResult = result;
      }
      goldenImageResult.parallel.columns[parallelism] = goldenParallelResult;
    });

    envGoldens[imageResult.name] = goldenImageResult;
  }

  if (ok) {
    test('Writing new goldens file', () {
      goldenValues.environments[env]!.images = envGoldens;
      golden.writeAsStringSync(
          JsonEncoder.withIndent('\t').convert(goldenValues));
    });
  }
}
