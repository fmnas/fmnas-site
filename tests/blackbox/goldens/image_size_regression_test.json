{
	"environments": {
		"panray_24CPU": {
			"images": {
				"puget.heic": {
					"name": "puget.heic",
					"size": "16.7 MB",
					"dimensions": "9248x6936",
					"parallel": {
						"columns": {
							"1": {
								"succeeded": 1,
								"failed": 0,
								"max": 2055,
								"total": 2055,
								"ram": 1860944
							},
							"2": {
								"succeeded": 2,
								"failed": 0,
								"max": 2236,
								"total": 4472,
								"ram": 2587356
							},
							"3": {
								"succeeded": 3,
								"failed": 0,
								"max": 2378,
								"total": 7133,
								"ram": 3284008
							}
						},
						"parallelLimit": 0
					},
					"group": null
				},
				"pigeons.jpg": {
					"name": "pigeons.jpg",
					"size": "12.1 MB",
					"dimensions": "4656x3492",
					"parallel": {
						"columns": {
							"1": {
								"succeeded": 1,
								"failed": 0,
								"max": 193,
								"total": 193,
								"ram": 1248172
							},
							"2": {
								"succeeded": 2,
								"failed": 0,
								"max": 210,
								"total": 420,
								"ram": 1336536
							},
							"3": {
								"succeeded": 3,
								"failed": 0,
								"max": 223,
								"total": 669,
								"ram": 1502376
							}
						},
						"parallelLimit": 0
					},
					"group": null
				},
				"train.heic": {
					"name": "train.heic",
					"size": "16.4 MB",
					"dimensions": "9248x6936",
					"parallel": {
						"columns": {
							"1": {
								"succeeded": 1,
								"failed": 0,
								"max": 2090,
								"total": 2090,
								"ram": 1792492
							},
							"2": {
								"succeeded": 2,
								"failed": 0,
								"max": 2245,
								"total": 4490,
								"ram": 2636416
							},
							"3": {
								"succeeded": 3,
								"failed": 0,
								"max": 2406,
								"total": 7217,
								"ram": 3381740
							}
						},
						"parallelLimit": 0
					},
					"group": null
				},
				"litter.jpg": {
					"name": "litter.jpg",
					"size": "12.6 MB",
					"dimensions": "8384x6035",
					"parallel": {
						"columns": {
							"1": {
								"succeeded": 1,
								"failed": 0,
								"max": 410,
								"total": 410,
								"ram": 1648040
							},
							"2": {
								"succeeded": 2,
								"failed": 0,
								"max": 448,
								"total": 896,
								"ram": 2083552
							},
							"3": {
								"succeeded": 3,
								"failed": 0,
								"max": 480,
								"total": 1439,
								"ram": 2495396
							}
						},
						"parallelLimit": 0
					},
					"group": null
				},
				"dubai.heic": {
					"name": "dubai.heic",
					"size": "18.6 MB",
					"dimensions": "9248x6936",
					"parallel": {
						"columns": {
							"1": {
								"succeeded": 1,
								"failed": 0,
								"max": 2168,
								"total": 2168,
								"ram": 1879708
							},
							"2": {
								"succeeded": 2,
								"failed": 0,
								"max": 2317,
								"total": 4634,
								"ram": 2742180
							},
							"3": {
								"succeeded": 3,
								"failed": 0,
								"max": 2470,
								"total": 7409,
								"ram": 3582968
							}
						},
						"parallelLimit": 0
					},
					"group": null
				},
				"george.jpg": {
					"name": "george.jpg",
					"size": "1.4 MB",
					"dimensions": "4322x3289",
					"parallel": {
						"columns": {
							"1": {
								"succeeded": 1,
								"failed": 0,
								"max": 104,
								"total": 104,
								"ram": 1175100
							},
							"2": {
								"succeeded": 2,
								"failed": 0,
								"max": 115,
								"total": 230,
								"ram": 1171476
							},
							"3": {
								"succeeded": 3,
								"failed": 0,
								"max": 129,
								"total": 387,
								"ram": 1394064
							}
						},
						"parallelLimit": 0
					},
					"group": null
				},
				"callie.jpg": {
					"name": "callie.jpg",
					"size": "352.6 KB",
					"dimensions": "800x600",
					"parallel": {
						"columns": {
							"1": {
								"succeeded": 1,
								"failed": 0,
								"max": 11,
								"total": 11,
								"ram": 1148012
							},
							"2": {
								"succeeded": 2,
								"failed": 0,
								"max": 18,
								"total": 35,
								"ram": 1127212
							},
							"3": {
								"succeeded": 3,
								"failed": 0,
								"max": 12,
								"total": 36,
								"ram": 1108780
							}
						},
						"parallelLimit": 0
					},
					"group": null
				}
			}
		}
	}
}
